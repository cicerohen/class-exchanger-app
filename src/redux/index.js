export * from './modules';
export * from './store';
