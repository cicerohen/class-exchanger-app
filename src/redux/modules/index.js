import * as ui from './ui';
import * as form from './form';
import * as firebase from './firebase';
import * as auth from './auth';
import * as router from './router';
import * as user from './user';

export {
  ui,
  form,
  firebase,
  auth,
  router,
  user,
};
