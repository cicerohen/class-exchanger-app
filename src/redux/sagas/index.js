import auth from './authSaga';
import firebase from './firebaseSaga';
import router from './routerSaga';
import user from './userSaga';
import exchange from './exchangeSaga';

export {
  auth,
  firebase,
  router,
  user,
  exchange,
};
