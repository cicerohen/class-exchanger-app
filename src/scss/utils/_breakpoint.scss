@function lower-bound($range) {
  @if length($range) <= 0 {
    @return 0;
  }
  @return nth($range,1);
}

@function upper-bound($range) {
  @if length($range) < 2 {
    @return 999999999999;
  }
  @return nth($range, 2);
}

$small-range:(0em, 39.9375em);
$medium-range:(40em, 63.9375em);
$large-range:(64em, 74.9375em);

$screen: "only screen" !default;
$small-only: "#{$screen} and (max-width: #{upper-bound($small-range)})" !default;
$medium-only: "#{$screen} and (min-width: #{lower-bound($medium-range)}) and (max-width: #{upper-bound($medium-range)})" !default;
$medium-up: "#{$screen} and (min-width: #{lower-bound($medium-range)})" !default;
$large-only: "#{$screen} and (min-width: #{lower-bound($large-range)}) and (max-width: #{upper-bound($large-range)})" !default;
$large-up: "#{$screen} and (min-width: #{lower-bound($large-range)})" !default;
