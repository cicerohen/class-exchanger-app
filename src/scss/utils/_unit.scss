@function strip-unit($num) {
  @return $num / ($num * 0 + 1);
}

@function rem-calc($value, $base: null) {
    
  @if $base == null {
    $base: 62.5%;
  }

  @if unit($base) == '%' {
    $base: ($base / 100%) * 16px;
  }

  @if type-of($value) != 'number' {
    @warn inspect($value) + ' was passed to rem-calc(), which is not a number.';
    @return $value;
  }

  @if unit($value) != 'rem' {
    $value: strip-unit($value) / strip-unit($base) * 1rem;
  }

  @if $value == 0rem {
    $value: 0;
  }
  
  @return $value;

}
