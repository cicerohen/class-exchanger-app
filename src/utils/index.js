import { Iterable, fromJS } from 'immutable';
import { createBrowserHistory } from 'history';

export const asImmutable = obj => (Iterable.isIterable(obj) ? obj : fromJS(obj));
export const emptyMap = asImmutable({});

export const buildActions = (moduleName, prefix, suffix) => ({
    REQUESTED: `${moduleName}/${prefix}_${suffix || ''}${suffix ? '_' : ''}REQUESTED`,
    STARTED: `${moduleName}/${prefix}_${suffix || ''}${suffix ? '_' : ''}STARTED`,
    SUCCEEDED: `${moduleName}/${prefix}_${suffix || ''}${suffix ? '_' : ''}SUCCEEDED`,
    FAILED: `${moduleName}/${prefix}_${suffix || ''}${suffix ? '_' : ''}FAILED`,
    FULFILLED: `${moduleName}/${prefix}_${suffix || ''}${suffix ? '_' : ''}FULFILLED`,
});

export const toArrayPath = path => path.split('/').filter(p => p.replace(' ', '').length > 0);

export const getHistory = () => {
    window.browserHistory = window.browserHistory || createBrowserHistory();
    return window.browserHistory;
};

export const isDevEnv = () => {
  const nodeEnv = process.env.NODE_ENV;
  return nodeEnv !== 'production' && nodeEnv !== 'test';
};
