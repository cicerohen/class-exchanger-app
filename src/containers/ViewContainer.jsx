import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { createStructuredSelector } from 'reselect';
import { ui } from '../redux';
import View from '../components/View';
import HeaderContainer from './HeaderContainer';
import SidebarContainer from './SidebarContainer';
import FooterContainer from './FooterContainer';

class ViewContainer extends PureComponent {
  closeSidebar = () => {
    const { closeSidebarAction } = this.props;
    closeSidebarAction();
  }

  renderHeader = () => {
    const { header } = this.props;
    return (header || <HeaderContainer />);
  }

  renderSidebar = () => {
    const { sidebar } = this.props;
    return (sidebar || <SidebarContainer />);
  }

  renderFooter = () => {
    const { footer } = this.props;
    return (footer || <FooterContainer />);
  };

  render() {
    const {
      sidebarIsOpen, children, header, sidebar, footer, ...others
    } = this.props;
    return (
      <View
        header={this.renderHeader()}
        sidebar={this.renderSidebar()}
        footer={this.renderFooter()}
        sidebarIsOpen={sidebarIsOpen}
        closeSidebarFn={this.closeSidebar}
        {...others}
      >
        {children}
      </View>
    );
  }
}

ViewContainer.propTypes = {
  /** optional */
  header: PropTypes.node,
  sidebar: PropTypes.node,
  footer: PropTypes.node,
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node),
  ]),
  /** required */
  sidebarIsOpen: PropTypes.bool.isRequired,
  closeSidebarAction: PropTypes.func.isRequired,
};

ViewContainer.defaultProps = {
  header: undefined,
  sidebar: undefined,
  footer: undefined,
  children: undefined,
};

const mapStateToProps = createStructuredSelector({
  sidebarIsOpen: ui.selectors.sidebarIsOpen,
});

const mapDispatchToProps = {
  closeSidebarAction: ui.actions.closeSidebar,
};

export default connect(mapStateToProps, mapDispatchToProps)(ViewContainer);
