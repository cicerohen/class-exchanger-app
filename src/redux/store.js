import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { combineReducers } from 'redux-immutable';
import createSagaMiddleware from 'redux-saga';
import { createLogger } from 'redux-logger';
import { asImmutable, isDevEnv } from '../utils';
import * as modules from './modules';
import * as allSagas from './sagas';

const getReducers = () => Object.keys(modules).reduce((acc, curr) => {
    const module = modules[curr];
    if (module.reducers && module.reducers.reducer) {
        acc[module.constants.default] = module.reducers.reducer;
    }
    return acc;
}, {});

export const createReduxStore = ({
  middlewares = [],
  initialState = {},
  enhancers = [],
} = {}) => {
  const sagaMiddleware = createSagaMiddleware();
  if (isDevEnv()) {
    middlewares.push(createLogger({
      collapsed: true,
    }));
  }
  middlewares.push(sagaMiddleware);
  enhancers.push(applyMiddleware(...middlewares));
  const enhancer = composeWithDevTools(...enhancers);
  const store = createStore(
    combineReducers(getReducers()),
    asImmutable(initialState),
    enhancer,
  );
  store.sagas = Object.keys(allSagas).map(sagaName => sagaMiddleware.run(allSagas[sagaName]));
  store.cancelSagas = () => store.sagas.forEach(saga => saga.cancel());
  return store;
};
