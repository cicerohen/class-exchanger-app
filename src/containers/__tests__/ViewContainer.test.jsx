import React from 'react';
import { shallow } from 'enzyme';
import ViewContainer from '../ViewContainer';
import { createReduxTestStore } from '../../utils/test';
import { ui } from '../../redux';

const minimalProps = {
  title: 'View Title',
};

describe('<ViewContainer /> Tests', () => {
  it('should render without exploding', (done) => {
    const testStore = createReduxTestStore();
    const wrapper = shallow(<ViewContainer store={testStore} />);
    expect(wrapper).toHaveLength(1);
    done();
  });

  it('should render correctly with minimal props', (done) => {
    const testStore = createReduxTestStore();
    const wrapper = shallow(
      <ViewContainer
        store={testStore}
        {...minimalProps}
      >
        <div>content</div>
      </ViewContainer>,
    );
    expect(wrapper.dive().shallow()).toMatchSnapshot();
    done();
  });

  it('should render correctly with header prop', (done) => {
    const testStore = createReduxTestStore();
    const wrapper = shallow(<ViewContainer
      store={testStore}
      header={
        <div className="header" />
        }
      {...minimalProps}
    />);
    expect(wrapper.dive().shallow()).toMatchSnapshot();
    done();
  });

  it('should render correctly with sidebar prop', (done) => {
    const testStore = createReduxTestStore();
    const wrapper = shallow(<ViewContainer
      store={testStore}
      sidebar={
        <div className="sidebar" />
        }
      {...minimalProps}
    />);
    expect(wrapper.dive().shallow()).toMatchSnapshot();
    done();
  });

  it('should render correctly with footer prop', (done) => {
    const testStore = createReduxTestStore();
    const wrapper = shallow(<ViewContainer
      store={testStore}
      footer={
        <div className="footer" />
        }
      {...minimalProps}
    />);
    expect(wrapper.dive().shallow()).toMatchSnapshot();
    done();
  });

  it('should render correctly when sidebar is opened', (done) => {
    const testStore = createReduxTestStore({
      configs: {
        initialState: {
          ui: {
            sidebarIsOpen: true,
          },
        },
      },
    });
    const wrapper = shallow(<ViewContainer
      store={testStore}
      sidebarIsOpen
      {...minimalProps}
    />);
    expect(wrapper.dive().shallow()).toMatchSnapshot();
    done();
  });

  it('should close sidebar when click on sidebar backdrop', (done) => {
    const testStore = createReduxTestStore({
      configs: {
        initialState: {
          ui: {
            sidebarIsOpen: true,
          },
        },
      },
    });
    const wrapper = shallow(<ViewContainer store={testStore} {...minimalProps} />);
    testStore.when(ui.actions.CLOSE_SIDEBAR, () => {
      const sidebarIsOpen = testStore.select(ui.selectors.sidebarIsOpen);
      expect(sidebarIsOpen).toBe(false);
      setImmediate(() => {
        wrapper.update();
        expect(wrapper.dive().shallow()).toMatchSnapshot();
        done();
      });
    });
    const sidebarBackdrop = wrapper.dive().shallow().find('.sidebarBackdrop');
    sidebarBackdrop.simulate('click');
  });
});
