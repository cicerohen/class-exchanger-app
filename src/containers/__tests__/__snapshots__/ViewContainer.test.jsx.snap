// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewContainer /> Tests should close sidebar when click on sidebar backdrop 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <div>
      <div
        className="headerContainer"
      >
        <Connect(HeaderContainer) />
      </div>
      <main
        className="mainContainer"
      >
        <h2
          className="viewTitle"
        >
          View Title
        </h2>
      </main>
      <div
        className="footerContainer"
      >
        <Connect(FooterContainer) />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ViewContainer /> Tests should render correctly when sidebar is opened 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <div
      className="sidebarRoot"
    >
      <div
        className="sidebarContainer"
      >
        <Connect(SidebarContainer) />
      </div>
      <div
        className="sidebarBackdrop"
        onClick={[Function]}
      />
    </div>
    <div>
      <div
        className="headerContainer"
      >
        <Connect(HeaderContainer) />
      </div>
      <main
        className="mainContainer"
      >
        <h2
          className="viewTitle"
        >
          View Title
        </h2>
      </main>
      <div
        className="footerContainer"
      >
        <Connect(FooterContainer) />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ViewContainer /> Tests should render correctly with footer prop 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <div>
      <div
        className="headerContainer"
      >
        <Connect(HeaderContainer) />
      </div>
      <main
        className="mainContainer"
      >
        <h2
          className="viewTitle"
        >
          View Title
        </h2>
      </main>
      <div
        className="footerContainer"
      >
        <div
          className="footer"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ViewContainer /> Tests should render correctly with header prop 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <div>
      <div
        className="headerContainer"
      >
        <div
          className="header"
        />
      </div>
      <main
        className="mainContainer"
      >
        <h2
          className="viewTitle"
        >
          View Title
        </h2>
      </main>
      <div
        className="footerContainer"
      >
        <Connect(FooterContainer) />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ViewContainer /> Tests should render correctly with minimal props 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <div>
      <div
        className="headerContainer"
      >
        <Connect(HeaderContainer) />
      </div>
      <main
        className="mainContainer"
      >
        <h2
          className="viewTitle"
        >
          View Title
        </h2>
        <div>
          content
        </div>
      </main>
      <div
        className="footerContainer"
      >
        <Connect(FooterContainer) />
      </div>
    </div>
  </div>
</div>
`;

exports[`<ViewContainer /> Tests should render correctly with sidebar prop 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <div>
      <div
        className="headerContainer"
      >
        <Connect(HeaderContainer) />
      </div>
      <main
        className="mainContainer"
      >
        <h2
          className="viewTitle"
        >
          View Title
        </h2>
      </main>
      <div
        className="footerContainer"
      >
        <Connect(FooterContainer) />
      </div>
    </div>
  </div>
</div>
`;
